algebird-core/src/main/scala/com/twitter/algebird/experimental/BloomFilter.scala

```diff
@@ -0,0 +1,403 @@
+
+    // Variable names correspond to those used in the paper.
+    val t = numBits
+    val n = sInverse(t).round.toInt
+    // Take the min and max because the probability formula assumes
+    // nl <= sInverse(t - 1) and sInverse(t + 1) <= nr
+    val nl =
+      scala.math.min(sInverse(t - 1).floor, (1 - approximationWidth) * n).toInt
+    val nr =
+      scala.math.max(sInverse(t + 1).ceil, (1 + approximationWidth) * n).toInt
+    val prob =
+      1 -
+        scala.math.exp(t - 1 - s(nl)) *
+          scala.math.pow(s(nl) / (t - 1), t - 1) -
+        scala.math.exp(-scala.math.pow(t + 1 - s(nr), 2) / (2 * s(nr)))
+
+    Approximate[Long](nl, n, nr, scala.math.max(0, prob))
+  }
+}
+
+/**
+ * Bloom Filter - a probabilistic data structure to test presence of an element.
+ *
+ * Operations
+ *   1) insert: hash the value k times, updating the bitfield at the index equal to each hashed value
+ *   2) query: hash the value k times.  If there are k collisions, then return true; otherwise false.
+ *
+ * http://en.wikipedia.org/wiki/Bloom_filter
+ */
+final case class BloomFilter[A](n: Int, w: Int)(implicit val hash: Hash128[A]) { self =>
+  object Hash extends Serializable {
+    final def apply(s: A): Array[Int] = {
+      val target = new Array[Int](n)
+      nextHash(s, 0, new Array[Int](4), 4, target)
+      target
+    }
+
+    private def splitLong(x: Long, buffer: Array[Int], idx: Int): Unit = {
+      // unfortunately, this is the function we committed to some time ago, and we have tests
+      // locking it down. x.toInt & 0x7fffffff should work, but this gives a few different values
+      def toNonNegativeInt(x: Long): Int =
+        (math
+          .abs(x)
+          .toInt) & 0x7fffffff // no change for positive numbers, converts Integer.MIN_VALUE to positive number
+
+      val upper = toNonNegativeInt(x >> 32)
+      val lower = toNonNegativeInt((x << 32) >> 32)
+      buffer(idx) = upper
+      buffer(idx + 1) = lower
+    }
+
+    @annotation.tailrec
+    private def nextHash(
+        valueToHash: A,
+        hashIndex: Int,
+        buffer: Array[Int],
+        bidx: Int,
+        target: Array[Int]
+    ): Unit =
+      if (hashIndex != n) {
+        val thisBidx = if (bidx > 3) {
+          val (a, b) = hash.hashWithSeed((n - hashIndex).toLong, valueToHash)
+          splitLong(a, buffer, 0)
+          splitLong(b, buffer, 2)
+          0
+        } else bidx
+
+        target(hashIndex) = buffer(thisBidx) % w
+        nextHash(valueToHash, hashIndex + 1, buffer, thisBidx + 1, target)
+      }
+  }
+
+  /**
+   * Bloom Filter data structure
+   */
+  sealed abstract class BF extends Serializable {
+    def numHashes: Int = n
+
+    def width: Int = w
+
+    /**
+     * The number of bits set to true in the bloom filter
+     */
+    def numBits: Int
+
+    /**
+     * Proportion of bits that are set to true.
+     */
+    final val density: Double = numBits.toDouble / width
+
+    def ++(other: BF): BF
```

Comment:
BTW: def |(that: BF): BF might be a nice API to add. It could be an alias for ++ but it is closer to what the Monoid does.